### DISCUSSION TITLE
Missing module in azure appservice

### DISCUSSION THREAD

**Discussion by ugursepetci** [vercel/next.js#56010] [App Router]
next I upgraded to 13.5.3. The build is running, but when the app service starts up, it gives cannot find module error logs. All pages give 500 errors.

We use a Windows machine when building.

```
Error: Cannot find module 'next/dist\client\components\static-generation-async-storage.external.js'
Require stack:
- /home/site/wwwroot/.next/server/app/page.js /home/site/wwwroot/node modules/next/dist/server/require.js /home/site/wwwroot/node modules/next/dist/server/next-server.js
- /home/site/wwwroot/node modules/next/dist/server/next.is
/home/site/wwwroot/node modules/next/dist/server/lib/start-server.js /home/site/wwwroot/node modules/next/dist/cli/next-start.is
/home/site/wwwroot/node modules/next/dist/lib/commands.js
- /home/site/wwwroot/node modules/next/dist/bin/next at Module. resolveFilename (node:internal/modules/cjs/loader: 1077:15)
at /home/site/wwwroot/node_modules/next/dist/server/require-hook.js:54:36 at Module. load (node:internal/modules/cjs/loader: 922:27)
at Module.require (node: internal/modules/cis/loader: 1143:19)
at Module.patchedRequire (/agents/nodejs/node modules/diagnostic-channel/dist/src/patchRequire.js:16:46)
at mod.require (/home/site/wwwroot/node_modules/next/dist/server/require-hook.js:64:28)
at require (node: internal/modules/cjs/helpers: 110:18)
at 5319 (/home/site/wwwroot/.next/server/app/page.js:1:614)
at t /home/site/wwwroot/.next/server/webpack-runtime.js:1:128)
at 1775 (/home/site/wwwroot/.next/server/chunks/8051.js:1:107062)
code: 'MODULE NOT FOUND' requireStack:
home/site/wwwroot/.next/server/app/page.js',
"home/site/wwwroot/node modules/next/dist/server/require.is' /hone/site/wwwroot/node modules/next/dist/server/next-server.js
" /home/site/wwwroot/node_modules/next/dist/server/next.js', home/site/wwwroot/node_modules/next/dist/server/lib/start-server.js', /home/site/wwwroot/node_modules/next/dist/cli/next-start.js', home/site/wwwroot/node modules/next/dist/lib/commands.js'
"
```

**Comment by raphourbe** [comment]
I upgraded today from Next 13.4.3 to Next 13.5.x (I tried with versions: 13.5.3, 13.5.2, 13.5.1), and got the same error. I am also using Windows to build my app.
It works rolling back to Next 13.4.19.

**Reply by OlexiiZhevakin** [reply]
How great are you
I have now lost the problem that the main data from mysql does not arrive on the site

**Reply by muriminjue** [reply]
This worked for me

**Reply by omarqra** [reply]
Next 13.4.19 didn't work for me.
it's give me _Aborted (core dumped)_

**Comment by omarqra** [comment]
I am facing the same problem, and i guess should return to previews versions, but i had problem with resource not being enough on v13.4.10, this #45508 .

**Reply by flynhigher** [reply]
I'm stuck with both issues, one in 13.4 and the other in 13.5.
Guess I have to downgrade to 13.2.

**Comment by Enngage** [comment]
Same problem here.

**Comment by Lucas-Bravo** [comment]
same here...

**Comment by jody3t** [comment]
same here

**Comment by LukasVitek** [comment]
same on linux

**Comment by DarkPromise** [comment]
Having the same error when deploying to AWS, downgrading to 13.4.19 fixed it for me.

**Reply by DarkPromise** [reply]
Just to give an update to anyone who comes here, the issue lies with the compiler mixing forward slashes and Posix-style slashes in the pages.js file. 

I fixed this by SSH into my EC2 instance, using vim to modify the page.js to change all forward slashes to backslashes and restarting the server. (This is a bad way to fix it, the compiler should automatically detect the system and set the slashes accordingly). As mentioned in a comment below by ericbergh, you have to compile in an actual linux machine (WSL won't work for some reason) in order to have correct paths. I eventually gave up and just dockerized my entire service which automatically setups the build correct.

Previously this issue didn't occur until SWC was added, so I assume the original hardcoded replacement code for slashes in 2 pull requests (one in 2018 and one in 2022) did not move over. (Anyway, the issue lies in that compiling with WSL on a windows machine still packages the production build with the windows version of swc, when it should be using linux's.

**Reply by FMP-Dev** [reply]
@ugursepetci,
Found the reason for the generation of incorrect paths during assembly.
When building the build, you need to fix the generation of directories by replacing 2 backslashes "\\" with 1 forward "/".
To do this, you need to make changes to the file: \node_modules\next\dist\build\handle-externals.js
replace string (180):
return `commonjs ${localRes.replace(/.*?next[/\\]dist/, "next/dist")}`;
on
return `commonjs ${localRes.replace(/.*?next[/\\]dist/, "next/dist").replace(/\\/g, "/")}`;

Suitable for versions 13.4.19++ 13.5+

* This is just a fix, but it's better if the developer takes this into account in the Next.JS build

@DarkPromise, thanks for the tip!

**Reply by shawnlwy85** [reply]
this work! on latest 13.5.6.

without change this Next Js 13.5.6 still unable to deploy on ubuntu server with node js,

same error as above happened again.
"Error: Cannot find module 'next/dist\client\components\static-generation-async-storage.external.js'"

I think NextJs might need to fix this, since 13.4.X already got this error.

Thanks very much !

**Comment by ericbergh** [comment]
We also ran into this issue. We could temporarily fix it by building on a Linux machine instead, but not ideal in the long run for us.

**Reply by DarkPromise** [reply]
Hello, I have tried building it on ubuntu via WSL, but it still did not work. What did you build it with if its working for yours?

**Reply by ericbergh** [reply]
I work at a large corp so I'm not directly responsible for the build pipeline. I know that we are running Ubuntu 20.x and Node 18.18.0 however

**Reply by DarkPromise** [reply]
Hi, thanks for the reply. I will try it out tomorrow at work. Thanks again.

**Comment by alessandrojcm** [comment]
Same issue here when compiled locally as standalone and served as a docker image. Using next 13.5.4 on Windows 11. Reverting to 13.4.19 fixed it for me as well.

**Reply by DarkPromise** [reply]
I fixed mine by compiling within docker instead of locally. If you have that option you can try it.

**Comment by almirjakubaitis** [comment]
same error on Linux  Ubuntu 20.04 - next 13.5.6.
works on windows 10

**Comment by TimFL** [comment]
This issue still persists when building on Windows. Building on Linux is currently not an option for us, is there no fix in sight?

**Reply by shawnlwy85** [reply]
https://github.com/vercel/next.js/discussions/56010#discussioncomment-7315581

The solution at above.

**Reply by shawnlwy85** [reply]
It work fine on me at next 13.5.6

**Comment by TimFL** [comment]
> It work fine on me at next 13.5.6
We ended up creating a makeshift runner pipeline to build the project on Linux. The comment above is merely a bandaid. I hope Next sees this and releases an official fix, because we can't realistically go in and patch this file every time we update.

**Comment by CodeMasterZeroOne** [comment]
Have this in Next 14.0.1 and the solution above has helped in deploying to Azure Linux container, Thank you for your help guys👍

**Comment by israellev** [comment]
Same happened to me at next version 13.5.6 and 14.0.1, I got this error only on google cloud (app engine) and not locally on windows 11,
after rolling back to next@13.4.19 it works fine

**Comment by alexop1000** [comment]
Since it seems no one have made an issue for this yet, I have gone ahead and done it: https://github.com/vercel/next.js/issues/58244
Let's hope it gets resolved soon.

**Comment by Francois-MUGOROZI** [comment]
I had the same problem with next 14, later I found that building my project in windows system and deploy to a Linux system lead to that issue.

I then built my project in Linux system and it worked [Link to my solution](https://stackoverflow.com/a/77894267/13555761)

**Comment by Painwraith** [comment]
I recommend not using CI/CD from Github to Azure App Service if you are experiencing this issue. Using default configurations, I successfully deployed my Next 14 project using Azure App Service through Visual Studio Code (Windows) Azure Plugin.

**Reply by tuke307** [reply]
local deployment always worked but not on azure with github ci/cd pipeline. i always encountered this error;
```
Error: Cannot find module '../server/require-hook'
```
deployment over visual studio code extensions for azure app service worked perfectly!!! thanks!

**Reply by tuke307** [reply]
did you found a solution with github ci/cd though?

**Reply by lmf-git** [reply]
Same issue but on AWS. I thought it was PM2's fault.

**Comment by echiu-infoway** [comment]
> did you found a solution with github ci/cd though?

potentially this might work https://stackoverflow.com/a/75580851

**Comment by rossdakin** [comment]
Please see: https://stackoverflow.com/a/79444948/14794
